File: numeric/primes500/primes.exs

defmodule Primes do
  def sieve_loop(n) do
    receive do
      {:test_candidate, from, candidate} ->
        send(from, {:tested, self(), rem(candidate, n) != 0})
        sieve_loop(n)
      {:stop} ->
        :ok
    end
  end

  def sieve([], _, _) do
    :ok
  end
  def sieve([head | tail], from, candidate) do
    send(head, {:test_candidate, from, candidate})
    sieve(tail, from, candidate)
  end

  def exit_sieve([]) do
    :ok
  end
  def exit_sieve([head | tail]) do
    send(head, {:stop})
    exit_sieve(tail)
  end

  def test_loop([], from) do
    send(from, {:is_prime})
  end
  def test_loop(sieve, from) do
    receive do
      {:tested, pid, true} ->
        test_loop(sieve -- [pid], from)
      {:tested, _, false} ->
        send(from, {:not_prime})
    end
  end

  def first_n(n, _, _, _) when n < 1 do
    []
  end
  def first_n(n, _, _, _) when n == 1 do
    [2]
  end
  def first_n(n, prior, sieve, _) when length(prior) >= n do
    exit_sieve(sieve)
    prior
  end
  def first_n(n, [], _, _) do
    sieve3 = spawn(fn -> sieve_loop(3) end)
    first_n(n, [2,3], [sieve3], 5)
  end
  def first_n(n, prior, sieve, candidate) do
    thisPid = self()
    tester = spawn(fn -> test_loop(sieve, thisPid) end)
    sieve(sieve, tester, candidate)
    receive do
      {:is_prime} ->
        newPid = spawn(fn -> sieve_loop(candidate) end)
        first_n(n, prior ++ [candidate], sieve ++ [newPid], candidate + 2)
      {:not_prime} ->
        first_n(n, prior, sieve, candidate + 2)
    end
  end

  def first_n(n, to) do
    send(to, {:print_primes, first_n(n, [], [], 2)})
  end
end

defmodule PrimesPrinter do
  def prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, []) do
    {p1, p2, p3, p4, p5, p6, p7, p8, p9, p10}
  end
  def prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, [head | tail])
    when length(p1) < 50 do
    prime_split_to_print(p1 ++ [head], p2, p3, p4, p5, p6, p7, p8, p9, p10, tail)
  end
  def prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, [head | tail])
    when length(p2) < 50 do
    prime_split_to_print(p1, p2 ++ [head], p3, p4, p5, p6, p7, p8, p9, p10, tail)
  end
  def prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, [head | tail])
    when length(p3) < 50 do
    prime_split_to_print(p1, p2, p3 ++ [head], p4, p5, p6, p7, p8, p9, p10, tail)
  end
  def prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, [head | tail])
    when length(p4) < 50 do
    prime_split_to_print(p1, p2, p3, p4 ++ [head], p5, p6, p7, p8, p9, p10, tail)
  end
  def prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, [head | tail])
    when length(p5) < 50 do
    prime_split_to_print(p1, p2, p3, p4, p5 ++ [head], p6, p7, p8, p9, p10, tail)
  end
  def prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, [head | tail])
    when length(p6) < 50 do
    prime_split_to_print(p1, p2, p3, p4, p5, p6 ++ [head], p7, p8, p9, p10, tail)
  end
  def prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, [head | tail])
    when length(p7) < 50 do
    prime_split_to_print(p1, p2, p3, p4, p5, p6, p7 ++ [head], p8, p9, p10, tail)
  end
  def prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, [head | tail])
    when length(p8) < 50 do
    prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8 ++ [head], p9, p10, tail)
  end
  def prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, [head | tail])
    when length(p9) < 50 do
    prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9 ++ [head], p10, tail)
  end
  def prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, [head | tail]) do
    prime_split_to_print(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10 ++ [head], tail)
  end
  def prime_split_to_print(primes) do
    prime_split_to_print([], [], [], [], [], [], [], [], [], [], primes)
  end

  def print_primes_lines([], [], [], [], [], [], [], [], [], []) do
    :ok
  end
  def print_primes_lines([head1 | tail1], [head2 | tail2], [head3 | tail3],
    [head4 | tail4], [head5 | tail5], [head6 | tail6], [head7 | tail7],
    [head8 | tail8], [head9 | tail9], [head10 | tail10]) do
      IO.write(:stdio, :io_lib.format(
        "    ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B~n",
        [head1, head2, head3, head4, head5, head6, head7, head8, head9, head10]))
      print_primes_lines(tail1, tail2, tail3, tail4, tail5, tail6,
        tail7, tail8, tail9, tail10)
  end
  def print_primes_lines(primes) do
    splitPrimes = prime_split_to_print(primes)
    print_primes_lines(elem(splitPrimes, 0), elem(splitPrimes, 1),
      elem(splitPrimes, 2), elem(splitPrimes, 3), elem(splitPrimes, 4),
      elem(splitPrimes, 5), elem(splitPrimes, 6), elem(splitPrimes, 7),
      elem(splitPrimes, 8), elem(splitPrimes, 9))
  end

  def print_primes(from) do
    receive do
      {:print_primes, primes} ->
        IO.puts("First Five Hundred Primes")
        print_primes_lines(primes)
        send(from, {:stop})
    end
  end
end

primaryPid = self()
printer = spawn(fn -> PrimesPrinter.print_primes(primaryPid) end)
spawn(fn -> Primes.first_n(500, printer) end)
receive do
  {:stop} -> :ok
end

File: numeric/primes500/primes.erl
-module(primes).
-export([main/0]).

sieveLoop(N) ->
    receive
        {'TEST', From, Candidate} ->
            From ! {'TESTED', self(), Candidate rem N /= 0},
            sieveLoop(N);
        {'EXIT', stop} ->
            ok
    end.

sieve([], _, _) -> ok;
sieve([H|T], From, Candidate) ->
    H ! {'TEST', From, Candidate},
    sieve(T, From, Candidate).

exitSieve([]) -> ok;
exitSieve([H|T]) ->
    H ! {'EXIT', stop},
    exitSieve(T).

testLoop([], From) -> From ! {'SUCCESS'};
testLoop(Sieve, From) ->
    receive
        {'TESTED', Pid, true} -> testLoop(Sieve -- [Pid], From);
        {'TESTED', _, false} -> From ! {'FAIL'}
    end.

first_n(N, _, _, _) when N < 1 -> [];
first_n(N, _, _, _) when N == 1 -> [2];
first_n(N, Prior, Sieve, _) when length(Prior) >= N ->
    exitSieve(Sieve),
    Prior;
first_n(N, [], _, _) ->
    Sieve3 = spawn(fun() -> sieveLoop(3) end),
    first_n(N, [2,3], [Sieve3], 5);
first_n(N, Prior, Sieve, Candidate) ->
    ThisPid = self(),
    Tester = spawn(fun() -> testLoop(Sieve, ThisPid) end),
    sieve(Sieve, Tester, Candidate),
    receive
        {'SUCCESS'} ->
            NewPid = spawn(fun() -> sieveLoop(Candidate) end),
            first_n(N, Prior ++ [Candidate], Sieve ++ [NewPid], Candidate + 2);
        {'FAIL'} ->
            first_n(N, Prior, Sieve, Candidate + 2)
    end.

firstN(N, To) -> To ! {first_n(N, [], [], 2)}.

primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, []) ->
    {P1, P2, P3, P4, P5, P6, P7, P8, P9, P10};
primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, [H | T])
    when length(P1) < 50 ->
    primeSplitToPrint(P1 ++ [H], P2, P3, P4, P5, P6, P7, P8, P9, P10, T);
primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, [H | T])
    when length(P2) < 50 ->
    primeSplitToPrint(P1, P2 ++ [H], P3, P4, P5, P6, P7, P8, P9, P10, T);
primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, [H | T])
    when length(P3) < 50 ->
    primeSplitToPrint(P1, P2, P3 ++ [H], P4, P5, P6, P7, P8, P9, P10, T);
primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, [H | T])
    when length(P4) < 50 ->
    primeSplitToPrint(P1, P2, P3, P4 ++ [H], P5, P6, P7, P8, P9, P10, T);
primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, [H | T])
    when length(P5) < 50 ->
    primeSplitToPrint(P1, P2, P3, P4, P5 ++ [H], P6, P7, P8, P9, P10, T);
primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, [H | T])
    when length(P6) < 50 ->
    primeSplitToPrint(P1, P2, P3, P4, P5, P6 ++ [H], P7, P8, P9, P10, T);
primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, [H | T])
    when length(P7) < 50 ->
    primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7 ++ [H], P8, P9, P10, T);
primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, [H | T])
    when length(P8) < 50 ->
    primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8 ++ [H], P9, P10, T);
primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, [H | T])
    when length(P9) < 50 ->
    primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9 ++ [H], P10, T);
primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, [H | T]) ->
    primeSplitToPrint(P1, P2, P3, P4, P5, P6, P7, P8, P9, P10 ++ [H], T).
primeSplitToPrint(Primes) ->
    primeSplitToPrint([], [], [], [], [], [], [], [], [], [], Primes).

printPrimesLines([], [], [], [], [], [], [], [], [], []) -> ok;
printPrimesLines([H1|T1], [H2|T2], [H3|T3], [H4|T4], [H5|T5],
    [H6|T6], [H7|T7], [H8|T8], [H9|T9], [H10|T10]) ->
    io:format("     ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B ~4..0B ~n",
        [H1, H2, H3, H4, H5, H6, H7, H8, H9, H10]),
    printPrimesLines(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10).
printPrimesLines(Primes) ->
    SplitPrimes = primeSplitToPrint(Primes),
    printPrimesLines(element(1, SplitPrimes), element(2, SplitPrimes),
        element(3, SplitPrimes), element(4, SplitPrimes), element(5, SplitPrimes),
        element(6, SplitPrimes), element(7, SplitPrimes), element(8, SplitPrimes),
        element(9, SplitPrimes), element(10, SplitPrimes)).

printPrimes(From) ->
    receive
        {Primes} ->
            io:format("First Five Hundred Primes~n"),
            printPrimesLines(Primes),
            From ! {'EXIT'}
    end.

main() ->
    Primary = self(),
    Printer = spawn(fun() -> printPrimes(Primary) end),
    spawn(fun() -> firstN(500, Printer) end),
    receive
        {'EXIT'} -> ok
    end.
